### PR TITLE
feat(commands): message clear command

### PR DESCRIPTION
**Pull request by almostSouji**
Introduce a command to clear messages in a provided range
- 1 message: clear up to and include this message
- 2 messages: clear in range between and including both messages
- 12h in the past max to safeguard (accidental) abuse
- Verbose confirmation flow

image

> **Warning**
> bulk delete logs can no longer assume a single author for all messages on bulk delete and need to be refactored to account for that

> **Note**
> Waiting for https://github.com/Naval-Base/yuudachi/pull/1080 to be merged, so this can build on the extended message logs when logging the clear as a separate action in guild logs.

### CODE REVIEW

File: packages/yuudachi/locales/en-US/translation.json [hunk 1]
```diff
@@ -13,7 +13,8 @@
 				"duration_format": "Wrong duration format.",
 				"use_autocomplete": "Could not resolve the provided option. Make sure to select an autocomplete option.",
 				"target_not_found": "The given member is not in this guild.",
-				"timed_out": "Action timer ran out."
+				"timed_out": "Action timer ran out.",
+				"not_message_id": "Provided value `{{- val}}` for argument `{{- arg}}` is not a valid message id."
```
Comment (iCrawl) [inline packages/yuudachi/locales/en-US/translation.json:17]:
```suggestion
				"not_message_id": "Provided value `{{val}}` for argument `{{arg}}` is not a valid message id."
```

No need for this.

File: packages/yuudachi/locales/en-US/translation.json [hunk 2]
```diff
@@ -283,6 +284,23 @@
...
+				},
+				"pending": "You are about to clear {{count}} message by $t(command.mod.clear.authors, {\"count\": {{author_count}}}) sent in {{time}}. Are you sure?",
+				"pending_plural": "You are about to clear {{count}} messages by $t(command.mod.clear.authors, {\"count\": {{author_count}}}) sent in {{time}}. Are you sure?",
+				"success": "Successfully cleared {{count}} messages by $t(command.mod.clear.authors, {\"count\": {{author_count}}}) sent in {{time}}",
+				"success_plural": "Successfully cleared {{count}} messages by $t(command.mod.clear.authors, {\"count\": {{author_count}}}) sent in {{time}}.",
+				"authors": "{{count}} author",
+				"authors_plural": "{{count}} authors",
+				"buttons": {
+					"execute": "Clear {{count}} message",
+					"execute_plural": "Clear {{count}} messages"
```
Comment (iCrawl) [inline packages/yuudachi/locales/en-US/translation.json:301]: This works a bit different in i18next: https://www.i18next.com/translation-function/plurals#singular-plural

File: packages/yuudachi/src/commands/moderation/clear.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
+import { ms } from '@naval-base/ms';
+import { ButtonStyle, CommandInteraction, ComponentType } from 'discord.js';
```
Comment (iCrawl) [inline packages/yuudachi/src/commands/moderation/clear.ts:2]:
```suggestion
import { ButtonStyle, type CommandInteraction, ComponentType } from 'discord.js';
```

File: packages/yuudachi/src/commands/moderation/clear.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+		locale: string,
+	): Promise<void> {
+		const reply = await interaction.deferReply({ ephemeral: true, fetchReply: true });
+
+		if (!/^\d{17,20}$/.test(args.last_message)) {
+			throw new Error(
+				i18next.t('command.common.errors.not_message_id', {
+					lng: locale,
+					val: args.last_message,
+					arg: 'from',
```
Comment (iCrawl) [inline packages/yuudachi/src/commands/moderation/clear.ts:26]:
```suggestion
					val: args.last_message,
					arg: 'from',
					lng: locale,
```

File: packages/yuudachi/src/commands/moderation/clear.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+				}),
+			);
+		}
+
+		if (args.first_message && !/^\d{17,20}$/.test(args.first_message)) {
+			throw new Error(
+				i18next.t('command.common.errors.not_message_id', {
+					lng: locale,
+					val: args.first_message,
+					arg: 'to',
```
Comment (iCrawl) [inline packages/yuudachi/src/commands/moderation/clear.ts:36]:
```suggestion
					val: args.first_message,
					arg: 'to',
					lng: locale,
```

File: packages/yuudachi/src/commands/moderation/clear.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+			);
+		}
+
+		const firstMessage = await interaction.channel!.messages.fetch(args.last_message).catch(() => {
+			throw new Error(
+				i18next.t('command.mod.clear.errors.no_message', {
+					lng: locale,
+					message_id: args.last_message,
+					// eslint-disable-next-line @typescript-eslint/no-base-to-string
+					channel: interaction.channel!.toString(),
```
Comment (iCrawl) [inline packages/yuudachi/src/commands/moderation/clear.ts:47]:
```suggestion
					message_id: args.last_message,
					// eslint-disable-next-line @typescript-eslint/no-base-to-string
					channel: interaction.channel!.toString(),
					lng: locale,
```

File: packages/yuudachi/src/commands/moderation/clear.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+			);
+		});
+		const lastMessage = args.first_message
+			? await interaction.channel!.messages.fetch(args.first_message).catch(() => {
+					throw new Error(
+						i18next.t('command.mod.clear.errors.no_message', {
+							lng: locale,
+							message_id: args.first_message,
+							// eslint-disable-next-line @typescript-eslint/no-base-to-string
+							channel: interaction.channel!.toString(),
```
Comment (iCrawl) [inline packages/yuudachi/src/commands/moderation/clear.ts:58]:
```suggestion
							message_id: args.first_message,
							// eslint-disable-next-line @typescript-eslint/no-base-to-string
							channel: interaction.channel!.toString(),
							lng: locale,
```

File: packages/yuudachi/src/commands/moderation/clear.ts [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+		}
+
+		const uniqueAuthors = new Set(messages.map((m) => m.author.id));
+		const latest = messages.first()!;
+		const earliest = messages.last()!;
+		const delta = latest.createdTimestamp - earliest.createdTimestamp;
+
+		const clearKey = nanoid();
+		const cancelKey = nanoid();
+		const clearButton = createButton({
```
Comment (iCrawl) [inline packages/yuudachi/src/commands/moderation/clear.ts:81]:
```suggestion
		const clearKey = nanoid();
		const cancelKey = nanoid();

		const clearButton = createButton({
```

File: packages/yuudachi/src/commands/moderation/clear.ts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+			customId: clearKey,
+			label: i18next.t('command.mod.clear.buttons.execute', { lng: locale, count: messages.size }),
```
Comment (iCrawl) [inline packages/yuudachi/src/commands/moderation/clear.ts:83]:
```suggestion
			label: i18next.t('command.mod.clear.buttons.execute', { count: messages.size, lng: locale }),
```

File: packages/yuudachi/src/functions/pruning/pruneMessages.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+	const earliestPossiblePrune = Date.now() - 12 * 60 * 60 * 1000;
+	while (
+		pivot ? oldest.createdTimestamp < pivot.createdTimestamp && pivot.createdTimestamp > earliestPossiblePrune : true
+	) {
+		console.log('fetching more', {
+			pivot: pivot?.createdTimestamp,
+			oldest: oldest.createdTimestamp,
+			res_size: res.size,
+		});
```
Comment (iCrawl) [inline packages/yuudachi/src/functions/pruning/pruneMessages.ts:44]: Intended?

File: packages/yuudachi/src/functions/pruning/pruneMessages.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+			if (message.createdTimestamp >= oldest.createdTimestamp && message.createdTimestamp >= earliestPossiblePrune) {
+				res.set(message.id, message);
+			} else {
+				break;
+			}
+		}
+		pivot = messages.last();
+	}
+
+	console.log('res size: ', res.size);
```
Comment (iCrawl) [inline packages/yuudachi/src/functions/pruning/pruneMessages.ts:61]: Same

File: packages/yuudachi/src/functions/pruning/pruneMessages.ts [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+	if (chunk.length) {
+		res.push([...chunk]);
+	}
+
+	return res;
+}
+
+export async function pruneMessages(channel: GuildTextBasedChannel, messages: Collection<Snowflake, Message>) {
+	console.log(messages.map((m) => `${m.author.tag}: ${m.content}`).join('\n'));
```
Comment (iCrawl) [inline packages/yuudachi/src/functions/pruning/pruneMessages.ts:85]: Same

**Review by iCrawl — CHANGES_REQUESTED** [review]

**Comment by almostSouji** [comment]
Waiting for https://github.com/Naval-Base/yuudachi/pull/1080 to be merged, so this can build on the extended message logs when logging the clear as a separate action in guild logs.

File: packages/yuudachi/src/commands/moderation/clear.ts [hunk 13]
```diff
@@ -0,0 +1,155 @@
+import { ms } from '@naval-base/ms';
+import { ButtonStyle, CommandInteraction, ComponentType } from 'discord.js';
```
Comment (iCrawl) [inline packages/yuudachi/src/commands/moderation/clear.ts:2]:
```suggestion
import { ButtonStyle, type CommandInteraction, ComponentType } from 'discord.js';
```

**Pull request merged**